sdk/database/dbplugin/v5/grpc_server.go

```diff
@@ -306,10 +310,22 @@ func (g *gRPCServer) Close(ctx context.Context, _ *proto.Empty) (*proto.Empty, e
 
 // Version forwards the version request to the underlying Database implementation.
 func (g *gRPCServer) Version(ctx context.Context, _ *logical.Empty) (*logical.VersionReply, error) {
-	impl, err := g.getDatabaseInternal(ctx)
+	impl, err := g.getOrCreateDatabase(ctx)
 	if err != nil {
-		return nil, err
+		// if this is called without a multiplexing context, like from the plugin catalog directly,
+		// then we won't have a database ID, so let's generate a new database instance
```

Comment:
Or to throw another option in there, maybe we could check if we expect to need an ID (`g.singleImpl == nil`), and if an ID is missing (`_, err := pluginutil.GetMultiplexIDFromContext(ctx); err != nil`), then pre-populate it if so? Thinking about it though, `getOrCreateDatabase()` feels like the more natural place to put that check, so maybe a combination of those two.

Reply:
The specific error it returns is `"unexpected number of IDs in metadata: (%d)"`, which isn't super helpful to try to catch, and is a little brittle.

I didn't put this in `getOrCreateDatabase` because it would be a legitimate bug, I think, if any method other than `Version()` was called without a multiplexing ID. The only reason `Version()` is called without one is because we are booting up the plugin just to check the version.

I also considered booting up the plugin in `singleImpl` / non-multiplexed mode just to read the version, but it seemed hacky.

I think out of these options, I would probably prefer to change the error to be the same every time and look for that error, and maybe move the logic out of `Version()` to a separate function (but not in `getOrCreateDatabase`)?

Reply:
Sounds reasonable to me. Just for my own understanding, how come we don't hit this issue with the type method? It must be because the specific database plugin doesn't need to implement that method, it's actually answered by the combined db engine right?

Reply:
The `Type()` function is not called to determine if a plugin is a database plugin or not: https://github.com/hashicorp/vault/blob/main/vault/plugin_catalog.go#L465

So, it doesn't come up there.